### ISSUE TITLE
Cheat Improvements

### ISSUE THREAD

**Issue by lasers** [libretro/RetroArch#3748]
- [ ] Improvement
* When loading a cheat file, print a notification too (to eliminate lack of user feedback).
`"Loading cheat Aladdin.cht"`

- [ ] Improvement
* When saving a custom cheat file, print a notification too (to eliminate lack of user feedback). `"Saving cheat Aladdin-lol.cht"`

- [ ] Improvement
* There is no option to delete custom saved cheats. I don't know enough about cheats/user interaction enough to acknowledge if we really need "Cheat File Save As"
as we could submit new/working cheats to RetroArch cheat database. In a way, this sounds like the manual cheatstate. I'm about to mention auto cheatstate next.

- [ ] Improvement
* When loading a game, load cheat too by default (if it exists). I assume everything is off by default in the cheat database. If off by default... When paused, you would be in Quick Menu, go to Cheats, the options would be already there. Toggle on your favorites and apply. Random examples.
  ```
  "Auto-loading cheat from "Aladdin (World).cht"
  ```

- [ ] Improvement
* If cheat code(s) applied, is it possible to make auto cheatstate for the game to load next time (if any) instead of cheats? Along with the loading notification for feedback... Random examples.
  ```
  "Auto-loading cheatstate from "Aladdin (World).cht.auto"
  ```

- [ ] Possible bug
* When changing a number on "Cheat Passes", the cheat menu will get reset. If I load a cheat and wanted to add something new, I would have to increase "Cheat Passes" first... and the menu just gets reset. I lost my cheat right away.

~~Added [CLOSED] https://github.com/libretro/RetroArch/issues/3709.~~ **CUSTOMIZED COPY&PASTE BELOW.**

- [ ] Possible bug 
* After loading a core and picking a random cheat, use "Cheat File Save As" to write a file "lol", you can try it again to overwrite the existing "lol" file without warning about overwriting an existing file.

- [ ] Possible bug
* After loading a core and picking a random cheat, use "Cheat FIle Save As" to write a file "lol", turn "Cheat Passes" down by 1 or more. Save as "lol" again. Craaash.

- [ ] Possible bug
* Remove column / scrolling text / compact mode to make it more readability without waiting on scrolling texts.

  ![2016-10-21__13 28 37_small](https://cloud.githubusercontent.com/assets/852504/19608936/daa6c70c-9799-11e6-8371-138e079896ca.png)

- [ ] Improvement
* Strip `.cht` from files or `(CHEAT)` from frontend? (or `.cht` from frontend)? When loading a cheat code, we see `.cht` and `(CHEAT)` in the list. Both means same thing. 

- [ ] History
* Remember last location and position when loading cheat... per core/system (using names from database) to easily switch cheats per core/system for the game.

- [X] Improvement
* **FIXED.** Rename `Cheat File Load` to `Load Cheat Code`. (Human)

- [X] Improvement
* **FIXED.** Rename `Cheat File Save As` to `Save Cheat Code As`. (Human)

- [X] Improvement
* **FIXED.** Rearrange options for better setup flow (imo) as `Apply Cheat Changes` comes between a changeable numbered list of cheats and `Cheat Passes` (number). Similar to options under  `Shader`.

  ```
  Apply Cheat Changes
  Load Cheat Code
  Save Cheat Code As
  Cheat Passes     3
  Cheat #0:        N/A: (OFF)
  Cheat #1:        N/A: (OFF)
  Cheat #2:        N/A: (OFF)
  ```
- [X] Possible bug
* **FIXED.** When loading a cheat file, the path on top-left corner will be displayed all the way merging through the text from cheat list.

- [ ] Improvement
* Display the filename too in cheat menu.
  
  ```
  Apply Cheat Changes
  Load Cheat Code
  Save Cheat Code As
  Cheat Code:      Christmas NiGHTS into Dreams...
  Cheat Passes     3
  Cheat #0:        N/A: (OFF)
  Cheat #1:        N/A: (OFF)
  Cheat #2:        N/A: (OFF)
  ```

  ```
  Apply Cheat Changes
  Load Cheat Code    Christmas NiGHTS into Dreams...
  Save Cheat Code As
  Cheat Passes     3
  Cheat #0:        N/A: (OFF)
  Cheat #1:        N/A: (OFF)
  Cheat #2:        N/A: (OFF)
  ```

- [ ] Improvement
* After customizing and saving a cheat code, RetroArch appends the name at the end after an input... like street fightr 4 (RetroArch).cht or its appropriate cheat device name.

**Comment by lasers** [comment]
Possible bug: After loading a core and picking a random cheat, use "Cheat File Save As" to write a file "lol", you can select it again to overwrite the existing "lol" file without any existing file warning.

Possible bug: After loading a core and picking a random cheat, use "Cheat FIle Save" to write a file "lol", turn "Cheat Passes" down by 1 or more. Save as "lol" again. Craaash.

**Comment by lasers** [comment]
Possible bug:
* Compact mode forever toggled on under cheats making it unpleasant to set up cheat options for no reason. Wild guess, alignment?

Readability:
* An option to turn on full text (aka no more scrolling text) for everything in menu, not just cheats.

Excessive:
* Strip `.cht` from files or from frontend? When loading a cheat code, we see `.cht` and `(CHEAT)` in the list. Both means same thing. 

History:
* Remember last location and position when loading cheat... per core/system (using names from database) to easily switch cheats per core/system for the game.

Improvements:
* Rename `Cheat File Load` to `Load Cheat Code`
* Rename `Cheat File Save As` to `Save Cheat Code As`
* Reorganize options for better setup flow (imo) as `Apply Cheat Changes` comes between a changeable numbered list of cheats and `Cheat Passes` (number). Similar to options under  `Shader`.
```
Apply Cheat Changes
Load Cheat Code
Save Cheat Code As
Cheat Passes     3
Cheat #0:        N/A: (OFF)
Cheat #1:        N/A: (OFF)
Cheat #2:        N/A: (OFF)
```

**Comment by lasers** [comment]
Possible bug: When loading a cheat file, the path on top-left corner will be displayed all the way through the cheat list.

**Comment by lasers** [comment]
UI Improvement: Display the filename too in cheat menu.
```
Apply Cheat Changes
Load Cheat Code
Save Cheat Code As
Cheat Code       Christmas NiGHTS into Dreams...
Cheat Passes     3
Cheat #0:        N/A: (OFF)
Cheat #1:        N/A: (OFF)
Cheat #2:        N/A: (OFF)
```

**Comment by lasers** [comment]
@twinaphex Would you (or somebody else) look into some of those? Start with turning off compact mode because the much longer cheat filenames (following No-Intro naming conviction) have been merged lately.

I think we might get some complaints about that in a week or two.

**Comment by lasers** [comment]
After customizing and saving a cheat code, RetroArch should append the name at the end after an input... like `street fightr 4 (RetroArch).cht`.
* Reason 1: Cheat codes and its formats are sometimes different.
* Reason 2: And it would just looks so retro. Enuff said!

**Comment by mleonid2000** [comment]
Hello!!!
I want to share my experience.
System and ROM: 
	Windows 7x64, retroarch 1.5.0(i'm test x64 and x86), core - genesis_plus_gx_libretro.dll. rom - Boogerman (U) [!].bin

	I'm load default cheat file "Boogerman - A Pick and Flick Adventure (USA, Europe) (Action Replay).cht"
Anything work fine, but in default cheat file does not have few cheat's.
I'm try to port cheat code from "Fusion 3.64" cheat's, but some cheat code does not work.
And i'm start to experiment:

Original file:
cheat0_desc = "Infinite Strong Belches/Burps/Farts"
cheat0_code = "FF023E:>25<"
cheat0_enable = false 

port file:
cheat2_desc = "Infinite Strong Belches/Burps/Farts"
cheat2_code = "FF023E:>0020<"
cheat2_enable = true

when i'm remove two zero everything is work.
But some cheat codes work with additionals zero, and some cheat's won't work without zeros.

Here full port of my cheat codes 100% working for me:

cheats = 7

cheat0_desc = "Infinite Lives"
cheat0_code = "FF022E:09"
cheat0_enable = true

cheat1_desc = "Infinite Hits"
cheat1_code = "FF1A3F:04"
cheat1_enable = true

cheat2_desc = "Infinite Strong Belches/Burps/Farts"
cheat2_code = "FF023E:20"
cheat2_enable = true

cheat3_desc = "Infinite Boogers"
cheat3_code = "FF023C:20"
cheat3_enable = true

cheat4_desc = "Milk boogers"
cheat4_code = "FF0240:0001"
cheat4_enable = true

cheat5_desc = "Pepper burp"
cheat5_code = "FF0241:0001" < That cheat code work only with additional zeros. ("FF0241:>01<") that combo does not work.
cheat5_enable = true

cheat6_desc = "Max Plungers"
cheat6_code = "FF0232:0050"
cheat6_enable = false

And i'm have question:
Which format I should use in the second part of the cheat code.("FF0241:>hex or dec<")?

p.s. some thing wrong with this zeros

Sorry if my post was useless.

**Comment by twinaphex** [comment]
@RetroSven Are there still parts in here that could serve as worthwhile feedback, or can this be closed by now?

**Comment by RetroSven** [comment]
I think pretty much all of that has been addressed or is no longer applicable.  Suggest closing.

@lasers please review the cheat system in the current RA and if there is still anything you feel needs addressing, please open a new ticket with those specific details.

**Comment by lasers** [comment]
Go ahead and close. I hadn't used RetroArch in a long time. Thank you.

**Comment by mleonid2000** [comment]
I need littlebit more time to check issue.(1-2 days). if it possible.

**Comment by twinaphex** [comment]
Alright, lets wait for your feedback.

**Comment by mleonid2000** [comment]
Thanks for waiting!
And now 
cheat6_desc = "Milk boogers"
cheat6_code = "FF0240:01" < That's ok! **But "FF0240:0001" is similar for "FF0241:01"** 
cheat6_enable = true

I think issue should be closed.
For now cheat engine for me work nice!!!
Thanks for payattention.

**Comment by orbea** [comment]
Both @RetroSven and @mleonid2000 suggested closing this issue and any further issues with the newer cheat system should be addressed in new issues.
